A grab bag of changes

randall opened the pull request:
new flag: --connect-to
    
    Allows user to override DNS for the initial config request. This is
    accomplished using a custom DialContext overring the hostname used to
    Dial to. This allows for TLS certificate validation to still
    happen(optionally) while connecting to TLS secured resources.
    
    Also,
        - allows for optional enforcement of certificate verification
        - stamp built git version into binary and adds a --version option
        - adds a user-agent to all outgoing request
        - exit(1) on failures for easier shell error detection
        - lint/deprecated function cleanup
        - output prometheus stats for use in a monitoring system
        - extended stats: fix build error on other platforms
        - allow for use of proxy through environment variables


new flag: --url
    
    This enables passing in a single URL for easier running
    
    eg networkQuality --url https://networkquality.example.com/path/to/config

new flag: --continue-on-conn-reuse
    
    Alternate behavior of instead of panicing, just continue the test run.

hawkinsw commented on rpm/rpm.go:
```diff
@@ -291,8 +290,9 @@ func CombinedProber(
 				)
 			}
-			transport := http2.Transport{}
+			transport := &http.Transport{}
```
I assume that you are comfortable that the use of the `http2.ConfigureTransport` call below will do the right thing and that we do not need to manually specify that we want HTTP2? It seems like your change is absolutely correct, I just want to verify that I am reading it correctly!

randall replied:
That's my understanding and I'm comfortable with it.

hawkinsw replied:
Great! Again, I agree and just wanted to make sure that we were on the same page!!

hawkinsw commented on rpm/rpm.go:
```diff
@@ -310,9 +310,15 @@ func CombinedProber(
+
+			if err := utilities.OverrideHostTransport(transport, foreignProbeConfiguration.ConnectToAddr); err != nil {
+				fmt.Printf("Fatal: Could not configure Transport: %s", err)
+				os.Exit(1)
```
I would prefer that these errors bubble up through the normal paths and that we let the caller decide what to do in the case of an error. Does that seem unreasonable?

randall replied:
I would too prefer that, but there's not a clear error exit point here.

Can you give guidance on your preference here? An error chan that's listened for?

hawkinsw replied:
Okay. I have thought lots about this and here's what I think we can easily do ... 

Let's commit this now and then in the next few days I will go through the code and build a centralized error-logging channel that we can distribute to the functions that could generate errors in certain conditions. If those functions generate non-critical errors (I would consider the above one such example because being unable to launch a *single* probe should not stop the entire test [although it would be different if we saw the error persistently, but that's an issue for the thing listening to the error channel (and now there are *so* many nested brackets [because of asides] that I forgot my train of thought)]) then they log errors to that channel. The receiver for messages on that channel can then decide how the overall test should proceed. 

Besides the strained English in this message, does that plan sound reasonable?

hawkinsw commented on utilities/transport.go:
```diff
@@ -0,0 +1,49 @@
+		return dialer.DialContext(ctx, network, addr)
+	}
+
+	if err := http2.ConfigureTransport(transport); err != nil {
```
Should we handle the case that this is not really an error:

> It returns an error if t1 has already been HTTP/2-enabled. 

That does not seem like a real "error" in this case.

randall replied:
So just ignore it?

hawkinsw replied:
It seems to me (according to the documentation) that the *only* error that could result is when HTTP/2 is already enabled. If you agree that is the case, then I would say, "yes, let's ignore it." I will defer to your judgment -- whatever way you decide is great.

randall replied:
Removed. And actually this fixes up the error path for this where now we don't need to bubble the error up. 2 birds, 1 stone.

hawkinsw replied:
Bingo! Thanks! 

I still think that we will want to have that "bubble-up path" but it's not as pressing now!

hawkinsw commented on networkQuality.go:
```diff
@@ -303,31 +355,35 @@ func main() {
+			URL:                config.Urls.LargeUrl,
+			KeyLogger:          sslKeyFileConcurrentWriter,
+			ConnectToAddr:      config.ConnectToAddr,
+			InsecureSkipVerify: config.InsecureSkipVerify,
```
I would prefer that we continue to refer here to `*insecureSkipVerify` to make it clear that the value comes from the command line and is not somehow overridden by the data that comes back from the server's configuration.

randall replied:
Done!

hawkinsw commented on networkQuality.go:
```diff
@@ -303,31 +355,35 @@ func main() {
+			URL:                config.Urls.UploadUrl,
+			KeyLogger:          sslKeyFileConcurrentWriter,
+			ConnectToAddr:      config.ConnectToAddr,
+			InsecureSkipVerify: config.InsecureSkipVerify,
```
See above.

hawkinsw commented on networkQuality.go:
```diff
@@ -303,31 +355,35 @@ func main() {
-			Host: config.Urls.SmallUrlHost,
+			URL:                config.Urls.SmallUrl,
+			ConnectToAddr:      config.ConnectToAddr,
+			InsecureSkipVerify: config.InsecureSkipVerify,
```
See above.

hawkinsw commented on networkQuality.go:
```diff
@@ -303,31 +355,35 @@ func main() {
-			Host: config.Urls.SmallUrlHost,
+			URL:                config.Urls.SmallUrl,
+			ConnectToAddr:      config.ConnectToAddr,
+			InsecureSkipVerify: config.InsecureSkipVerify,
```
See above.

hawkinsw commented on lgc/lgc.go:
```diff
@@ -356,29 +355,30 @@ func (lgd *LoadGeneratingConnectionDownload) doDownload(ctx context.Context) {
-	transferred := atomic.SwapUint64(&lgd.uploaded, 0)
-	newIntervalEnd := (time.Now().Sub(lgd.uploadStartTime)).Nanoseconds()
-	previousIntervalEnd := atomic.SwapInt64(&lgd.lastIntervalEnd, newIntervalEnd)
+func (lgu *LoadGeneratingConnectionUpload) TransferredInInterval() (uint64, time.Duration) {
```
Thank you for catching this!!

hawkinsw reviewed (changes requested):
As I said on slack, these changes are great. I just had a few questions. Thanks!!

randall commented:
FYI I've modified Darwin.go under extended stats (fixing and stylistic changes): https://github.com/network-quality/goresponsiveness/pull/44/commits/8a66c0ad439a6a5d2ae886b28ec841b3663df226

and I've added a new flag (--continue-on-conn-reuse) to make the panic-ing when connection reuse is detected optional.

hawkinsw commented on rpm/rpm.go:
```diff
@@ -197,7 +199,9 @@ func Probe(
 	if (probeType == SelfUp || probeType == SelfDown) && !probeTracer.stats.ConnectionReused {
-		panic(!probeTracer.stats.ConnectionReused)
+		if !continueOnConnReuse {
+			panic(!probeTracer.stats.ConnectionReused)
```
I am conflicted about this. Having a `panic` here is (I think) a very important part of the test because of the semantics of the protocol. It is a requirement that the self probes execute on a pre-existing connection. If that is not the case then it seems like the integrity of the entire test is up for debate.

What do you think?

randall replied:
Yeah, after sleeping on it a bit, I agree. I'm going to drop this particular change from the patch set.

hawkinsw replied:
Perfect! When we have the consolidated error-reporting feature, we can remove the `panic` and make it an error that the listener can deal with!

hawkinsw reviewed (changes requested):
Thanks again -- I think that we are ready to go when you are (given the few comments that I made are basically just rhetorical questions). Thanks for all these changes! They are great!
